### DISCUSSION TITLE
Crashes; How can I make Crazyflies more stable during flight?

### DISCUSSION THREAD

**Discussion by imreburn** [bitcraze/discussions#937] [Q&A] [answered]
Hi,

I am experiencing crashes of crazyflies (CF) and would like to make them more stable and reliable in terms of hardware. 

### Set-up
CFs were updated (firmware version 2023.07). I am using two lighthouse V2 beacons, and I think that setting them up is done properly (all green). The localization accuracy looks good when manually moving CFs, although CFs sometimes lose lighthouse signals (orange LED) where I believe they are supposed to receive signals well. Sunlight and reflections are blocked.
### Issue
CFs sometimes crash soon after take-off or during mission. Crashes just after take-off are more common. So far, I have run only simple missions included in the examples. Sometimes propellers are flying off. 
### Analysis
I think it is related with hardware (propellers, motors) or assembly . Most of crashes occurred just after take-off. Furthermore, the ones that did not crash in the first flight are usually doing good for the rest of experiments. 
### What I did
1. I saw a [post](https://github.com/orgs/bitcraze/discussions/141) that recommends hotgluing propeller in order to firmly adhere to the motor shaft. I didn't try it yet, but I just test each CF by take-off while holding it by hand to see if propellers are not escaping.
2. I check propeller balance.
3. I found "propeller test" in the client. I am not sure what exactly the test executes and checks, but it seems that less crash happened with less variance.

It got a bit improved with 1-3. But I want to reduce crashes as much as I can, since damages can be accumulated by each crash even if CF looks okay.
### Questions
1. Propellers
  - I found new 47-17 propellers and HQProp propellers included in the thrust upgrade kit. Can they help improve the stability?
  - I try to level four propellers from ground or body. Can this help? Sometimes propellers easily go all the way to the motors, leaving no gap between them. 
2. Propeller test: Can you provide how I can improve the variance? Now I just assemble again and again to see any difference.
3. In terms of balance, would it be better to fix the battery and decks? For me, fixing battery may not be an option since they should be stored separately.
4. Are there any other hardware checkpoints?
5. Are there any tests that I can check stability without actual take-off?

I would greatly appreciate if you have any comments or advice.

Thanks for reading!

**Comment by knmcguire** [comment, marked as answer]
Hi! Thanks for the very clear question and things you tried before. This will make it much easier for us to help you with this!

Let me just first clarify the propeller test. If there is more variance, that means that the propeller is causing vibration and it has been bent somehow (due to crashes). Indeed replacing that propeller with a new one helps quite a lot with the stability. Mind that a high vibration might also indicate that the motor shaft is bend.

Now about your questions:

1. these new propellers, the [High Quality props](https://store.bitcraze.io/collections/accessories/products/hqprop-51mmx2-1mm-light-grey--2cw-2ccw), and the new 47-17 propellers have an large improvement of efficiency, which means that the propeller generates more trust with the same spin that the regular props that come with the crazyflie. This will improve stability, as with fast movements, the small motors get saturated quite quickly. This tend to happen if you take off with a setpoint and not using the high level commander's take off. The crazyflie's motors will try to go there as quickly as they can, but then one or 2 motors have hit their limit, and then a crash occures. This happens way less often with more efficient propellers or [better motors](https://store.bitcraze.io/products/4-x-7x20-mm-dc-motor-pack-for-crazyflie-2-x-thrust-upgrade). Just be aware that the HQ props only fit the better motors so you will need to buy them as a bundle. The 47-17 fits on the regular motors that comees with the crazyflie. And About the leveling the props, it can help a bit but haven't seen this really making that much of a difference to be honest. But we haven't tested this thoroughly either. Just make sure that they are not at the very end as you will have an higher chance of them flying off. 

2. The variance can only be fixed by either the instructions of [balancing propellers with tape](https://www.bitcraze.io/documentation/tutorials/balancing-propellers/), or by replacing it with a new one. Also, remember what I told you earlier about the shaft being potentially bent so make sure to check for that as well.
3. In terms of balance, there might be a shift happening during the takeoff of the batteries, but just making sure that the battery is fixed by the decks should be enough (we haven't noticed this being an issue). Or else you can maybe consider a rubber band? During our demos we sometimes also[ solder the decks directly to the pins](https://www.bitcraze.io/documentation/tutorials/solder-deck-pins/) but that is a bit of a drastic measure maybe. Sometimes with landing, the deck tends to shift off if they have been placed on and off the crazyflie too many times (then the female pins on the deck have too much slack. In this case a small rubber band around the decks, or using [larger pins](https://store.bitcraze.io/products/male-long-deck-connector) will also do the trick.
4. Battery testing is important too. If your battery is bad but you want to do a quick take-off, this will not be able to provide the power to the motors adequately. For battery maintenance, it is important to not drain the battery completely such that it falls into the air. Also make sure that the motor mounts are not bent in the wrong direction, and that the motors are in good condition too (the variance prop test can help with that)
5. Software wise, yes. We have several example scripts that look at the [kalman estimator variance](https://github.com/bitcraze/crazyflie-lib-python/blob/f35b0baf6ff243d618f7a8becf5a69b1ec6c1f56/examples/autonomy/autonomous_sequence_high_level.py#L65) of the position before take-off (although just looking at the position variance might already be enough). Also, as I mentioned before, too quick take-offs are where most crashes happen, so either reduce the speed or use the [high level commander](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/high_level_commander/)'s take off function. The latter uses a trajectory onboard that starts off slow. Moreover, you can also check your [controller parameter](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#stabilizercontroller). PID is perhaps not as snappy as Mellinger but it is way more stable (see [this documentation explaining controllers](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/controllers/))


That was a long answer! I hope this was able to give you some tips to improve the stability! How many Crazyflies are you flying with at the same time?

**Reply by imreburn** [reply]
Hi, I really appreciate your advice that answered every single points I brought and included many resources. I have tried 4 CFs so far with Crazyradio, and aim to operate 8 CFs without radio (decentralized). I think all points above are important and I will check those. Especially it is valuable to know about take-off speed. There are actually a few ways to maneuver CFs explained in the docs and examples, and I was a bit curious about differences. The ones that I used - `autonomousSequence.py` or `swarmSequence.py` - do not use high level commander. I also think reducing the speed is a good point - is there a way to limit the speed of motor globally via lib(or firmware)?

**Reply by knmcguire** [reply]
Hi! I'm sorry, I must have missed your last question since you selected the answer already. 

So there are quite some examples in the [autonomy example folder](https://github.com/bitcraze/crazyflie-lib-python/tree/master/examples/autonomy). You can try out the autonomous_sequence_high_level.py, which has a take off example (you can comment out the trajectory if you just want to try the take off). There are also some automated api documentation to be found here: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/high_level_commander/

And yes you are able to reduce the maximum velocity, but that is done per controller. This is the parameter for the position PID controller: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#posctlpidzvelmax.

If you have any other questions, probably best to start up a new discussion :)
